File: src/programs/List.hs
module List where

import Syntax
import Num
import Bool
import Prelude hiding (succ)
import Text.Printf

-- Tests
infixr 9 %

nil :: Term a
nil = C "Nil" []

(%) :: Term a -> Term a -> Term a
x % y = C "Cons" [x, y]

lit :: Name -> Term a
lit x = C x []

a :: Term a
a = lit "a"

b :: Term a
b = lit "b"

c :: Term a
c = lit "c"

d :: Term a
d = lit "d"

list :: Show a => Term a -> String
list (V n) = printf "._%s" (show n)
list (C "Cons" [h, t]) = printf "%s %% %s" (list h) (list t)
list (C "Nil"  _     ) = "nil"
list (C s []) = s
list x = show x

listo :: G a -> G a
listo g =
  let x = V "x" in
  let h = V "h" in
  let t = V "t" in
  Let (def "listo" ["x"] (( x === nil ) ||| (fresh ["h", "t"] (x === h % t &&& call "listo" [t])))
      ) g

membero :: G a -> G a
membero g =
  let x = V "x" in
  let list = V "list" in
  let h = V "h" in
  let t = V "t" in
  Let ( def "membero" ["x", "list"]
        (
          fresh ["h", "t"] ( ( list === h % t ) &&&
                             ( ( x === h ) |||
                               ( call "membero" [x, t])
                             )
                           )
        )
      ) g

inBotho :: G a -> G a
inBotho g =
  let x = V "x" in
  let ys = V "ys" in
  let zs = V "zs" in
  Let ( def "inBotho" ["x", "ys", "zs"] ( call "membero" [x, ys] &&& call "membero" [x, zs]) ) $ membero g

nilo :: G a -> G a
nilo g =
  let l = V "l" in
  Let ( def "nilo" ["l"] ( l === nil ) ) g

singletono :: G a -> G a
singletono g =
  let l = V "l" in
  let x = V "x" in
  Let ( def "singletono" ["l", "x"] ( l === x % nil ) ) g

maxLengtho :: G a -> G a
maxLengtho g =
  Let (def "maxLengtho" ["x", "m", "l"] (call "maxo" [x, m] &&& call "lengtho" [x, l])) $ maxo $ lengtho g
  where
    x = V "x"
    m = V "m"
    l = V "l"

copy :: G a -> G a
copy g =
  Let (def "copy" ["l", "c"]
        (
          (l === nil &&& c === nil) |||
          (fresh ["h", "t", "t'"]
            (
              l === h % t &&& c === h % t' &&& call "copy" [t, t']
            )
          )
        )
      ) g
    where
      l = V "l"
      c = V "c"
      h = V "h"
      t = V "t"
      t' = V "t'"

copy2 :: G a -> G a
copy2 g =
  Let (def "copy2" ["l", "c"]
        (
          (l === nil &&& c === nil) |||
          (fresh ["h"] (l === h % nil &&& c === h % nil)) |||
          (fresh ["h1", "h2", "t", "t'"]
            (
              l === h1 % (h2 % t) &&& c === h1 % t' &&& call "copy2" [t, t']
            )
          )
        )
      ) g
    where
      l = V "l"
      c = V "c"
      h = V "h"
      h1 = V "h1"
      h2 = V "h2"
      t = V "t"
      t' = V "t'"

copycopy :: G a -> G a
copycopy g =
  Let (def "copycopy" ["l", "l1", "l2"]
        (call "copy" [l, l1] &&& call "copy2" [l, l2])
      ) $ copy $ copy2 g
    where
      l = V "l"
      l1 = V "l1"
      l2 = V "l2"



lengtho :: G a -> G a
lengtho g =
  let x = V "x" in
  let l = V "l" in
  let h = V "h" in
  let t = V "t" in
  let z = V "z" in
  Let (def "lengtho" ["x", "l"]
        (
          (x === nil &&& l === zero) |||
           fresh ["h", "t", "z"]
             (x === h % t &&& l === succ z &&& call "lengtho" [t, z])
        )
      ) g

maxo :: G a -> G a
maxo g =
  Let (def "maxo" ["x", "m"] (call "maxo1" [x, zero, m])) $ maxo1 g
  where
    maxo1 g =
      Let (def "maxo1" ["x", "n", "m" ]
            (
              (x === nil &&& m === n) |||
              fresh ["h", "t", "z"] (x === h % t &&& call "leo" [h, n, trueo]  &&& call "maxo1" [t, n, m]) |||
              fresh ["h", "t", "z"] (x === h % t &&& call "gto" [h, n, trueo]  &&& call "maxo1" [t, h, m])
            )
          ) $ leo $ gto g
    x = V "x"
    m = V "m"
    n = V "n"
    h = V "h"
    t = V "t"
    z = V "z"

appLengtho :: G a -> G a 
appLengtho g = 
  Let (
    def "appLengtho" [] (
      fresh ["xs", "ys", "zs", "m", "n", "s"] 
        (
          call "appendo" [xs, ys, zs] &&& 
          call "lengtho" [xs, m] &&& 
          call "lengtho" [ys, n] &&& 
          call "lengtho" [zs, s] &&& 
          call "addo" [m, n, s]
        ) 
    )
  ) $ addo $ lengtho $ appendo g  
    where [xs, ys, zs, m, n, s] = map V ["xs", "ys", "zs", "m", "n", "s"]

appendo :: G a -> G a
appendo g =
  let x  = V "x"  in
  let y  = V "y"  in
  let xy = V "xy" in
  let h  = V "h"  in
  let t  = V "t"  in
  let ty = V "ty" in
  Let
    (def "appendo" ["x", "y", "xy"]
         ((x === nil &&& xy === y) |||
           fresh ["h", "t", "ty"]
             (x === h % t &&& xy === h % ty &&& call "appendo" [t, y, ty])
         )
    ) g

appendo' :: G a -> G a
appendo' g =
  let x  = V "x"  in
  let y  = V "y"  in
  let xy = V "xy" in
  let h  = V "h"  in
  let t  = V "t"  in
  let ty = V "ty" in
  Let
    (def "appendo'" ["x", "y", "xy"]
           ((x === nil ||| xy === y) |||
             fresh ["h", "t", "ty"]
               (x === h % t ||| xy === h % ty ||| call "appendo'" [t, y, ty])
           )
    ) g

reverso :: G a -> G a
reverso g =
  let x  = V "x"  in
  let y  = V "y"  in
  let h  = V "h"  in
  let t  = V "t"  in
  let rt = V "rt" in
  Let
    (def "reverso" ["x", "y"]
           ((x === nil &&& y === nil) |||
             fresh ["h", "t", "rt"]
               (x === h % t &&&
                 call "reverso" [t, rt] &&& call "appendo" [rt, h % nil, y])
           )
    ) $ appendo g

doubleReverso :: G a -> G a 
doubleReverso g = 
  let xs = V "xs" in 
  let sx = V "sx" in 
  Let 
    (def "doubleReverso" ["xs"]
      (fresh ["sx"] (call "reverso" [xs, sx] &&& call "reverso" [sx, xs]))
    ) $ reverso g 

revAcco :: G a -> G a
revAcco g =
  let xs = V "xs"
      acc = V "acc"
      sx = V "sx"
      h = V "h"
      t = V "t"
  in
  Let
    (def "revacco" ["xs", "acc", "sx"]
       (
         (xs === nil &&& sx === acc) |||
         (fresh ["h", "t"]
           (xs === h % t) &&&
           call "revacco" [t, h % acc, sx]
         )
       )
    ) g


assoco g = 
  Let (def "assoco" ["x", "xs", "v"] (
    fresh ["a", "b", "tl"] (
      xs === (C "pair" [a, b]) % tl &&& 
      (a === x &&& b === v ||| 
       call "assoco" [x, tl, v] -- &&& a =/= x
      )
  ))) g
    where [x, xs, v, a, b, tl] = map V ["x", "xs", "v", "a", "b", "tl"]

  -- let rec assoco x xs v =
  --   Fresh.three (fun a b tl ->
  --     (xs === (LPair.pair a b) % tl) &&&
  --     conde [
  --       (a === x) &&& (b === v);
  --       (a =/= x) &&& (assoco x tl v)
  --     ]
  --   )

File: test/resources/newSyntax/list.pl
fail() :- fail().
listo(nil).
listo(cons(H, T)) :- listo(T).
membero(H, cons(H, T)).
membero(X, cons(H, T)) :- membero(X, T).
inBotho(X, Ys, Zs) :- membero(X, Ys), membero(X, Zs).
nilo(nil).
singletono(cons(X, nil), X).
maxLengtho(X, M, L) :- maxo(X, M), lengtho(X, L).
maxMino(X, M, L) :- maxo(X, M), mino(X, L).
copy(nil, nil).
copy(cons(H, T), cons(H, T_0)) :- copy(T, T_0).
copy2(nil, nil).
copy2(cons(H, nil), cons(H, nil)).
copy2(cons(H1, cons(H2, T)), cons(H1, T_0)) :- copy2(T, T_0).
copycopy(L, L1, L2) :- copy(L, L1), copy2(L, L2).
lengtho(nil, zero).
lengtho(cons(H, T), succ(Z)) :- lengtho(T, Z).
lengtho_0(nil, zero).
lengtho_0(cons(H, T), succ(Z)) :- lengtho_0(T, Z).
maxo(X, M) :- maxo1(X, zero, M).
maxo1(nil, N, N).
maxo1(cons(H, T), N, M) :- leo(H, N, trueo), maxo1(T, N, M).
maxo1(cons(H, T), N, M) :- gto(H, N, trueo), maxo1(T, H, M).
mino(nil, zero).
mino(cons(H, T), M) :- mino1(T, H, M).
mino1(nil, N, N).
mino1(cons(H, T), N, M) :- leo(H, N, trueo), mino1(T, H, M).
mino1(cons(H, T), N, M) :- gto(H, N, trueo), mino1(T, N, M).
appLengtho() :- appendo(Xs, Ys, Zs), lengtho(Xs, M), lengtho(Ys, N), lengtho(Zs, S), addo(M, N, S).
appendo(nil, Xy, Xy).
appendo(cons(H, T), Y, cons(H, Ty)) :- appendo(T, Y, Ty).
appendo_0(nil, Y, Y).
appendo_0(cons(H, T), Y, Xy).
appendo_0(X, Y, cons(H, Ty)).
appendo_0(X, Y, Xy) :- appendo_0(T, Y, Ty).
reverso(nil, nil).
reverso(cons(H, T), Y) :- reverso(T, Rt), appendo(Rt, cons(H, nil), Y).
doubleReverso(Xs) :- reverso(Xs, Sx), reverso(Sx, Xs).
revAcco(nil, Acc, Acc).
revAcco(cons(H, T), Acc, Sx) :- revacco(T, cons(H, Acc), Sx).
assoco(X, cons(pair(X, V), Tl), V).
assoco(X, cons(pair(A, B), Tl), V) :- assoco(X, Tl, V).
nthOpt(nil, N, none).
nthOpt(cons(H, T), zero, some(H)).
nthOpt(cons(H, T), succ(X), R) :- nthOpt(T, succ(X), R).
fail() :- fail().
notZero(s(Y)).
addo(o, Z, Z).
addo(s(X_0), Y, s(Z_0)) :- addo(X_0, Y, Z_0).
mulo(o, Y, o).
mulo(s(X_0), Y, Z) :- addo(Y, Z_0, Z), mulo(X_0, Y, Z_0).
leo(o, Y, trueo).
leo(s(Zz), o, falso).
leo(s(X_0), s(Y_0), B) :- leo(X_0, Y_0, B).
gto(s(Zz), o, trueo).
gto(o, Y, falso).
gto(s(X_0), s(Y_0), B) :- gto(X_0, Y_0, B).
geo(X, Y, Z) :- leo(Y, X, Z).
lto(X, Y, Z) :- gto(Y, X, Z).